go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -300,28 +301,58 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *corev1.Service, netI
 					var nodeportFlows []string
-					klog.V(5).Infof("Adding flows on breth0 for Nodeport Service %s in Namespace: %s since ExternalTrafficPolicy=local", service.Name, service.Namespace)
+					klog.V(5).Infof("Adding flows on breth0 for Nodeport Service %s with targetPort %d in "+
+						"Namespace: %s since ExternalTrafficPolicy=local", service.Name, targetPort, service.Namespace)
 					// table 0, This rule matches on all traffic with dst port == NodePort, DNAT's the nodePort to the svc targetPort
-					// If ipv6 make sure to choose the ipv6 node address for rule
+					// If IPv6 make sure to choose the IPv6 node address for rule, otherwise send to the IPv4 node address.
+					gatewayAddress := npw.gatewayIPv4
 					if strings.Contains(flowProtocol, "6") {
-						nodeportFlows = append(nodeportFlows,
-							fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=[%s]:%s),table=6)",
-								cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort, config.Default.HostNodePortConntrackZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
-					} else {
-						nodeportFlows = append(nodeportFlows,
-							fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
-								cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort, config.Default.HostNodePortConntrackZone, npw.gatewayIPv4, svcPort.TargetPort.String()))
+						gatewayAddress = fmt.Sprintf("[%s]", npw.gatewayIPv6)
+					}
+					// This is just an additional safety check - we should never get into this situation.
+					// If gatewayAddress is empty, we cannot install the following flow and we will continue with the
+					// next flowProtocol.
+					if gatewayAddress == "" || gatewayAddress == "[]" {
+						err := fmt.Errorf("gateway IP not available for protocol %s", flowProtocol)
+						errors = append(errors, err)
+						continue
 					}
+
+					nodeportFlows = append(nodeportFlows,
+						fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, "+
+							"actions=ct(commit,zone=%d,nat(dst=%s:%d),table=6)",
+							cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort,
+							config.Default.HostNodePortConntrackZone, gatewayAddress, targetPort))
+
 					nodeportFlows = append(nodeportFlows,
 						// table 6, Sends the packet to the host. Note that the constant etp svc cookie is used since this flow would be
 						// same for all such services.
 						fmt.Sprintf("cookie=%s, priority=110, table=6, actions=output:LOCAL",
 							etpSvcOpenFlowCookie),
 						// table 0, Matches on return traffic, i.e traffic coming from the host networked pod's port, and unDNATs
-						fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%s, actions=ct(zone=%d nat,table=7)",
-							cookie, flowProtocol, svcPort.TargetPort.String(), config.Default.HostNodePortConntrackZone),
+						// Return traffic should not be committed to conntrack.
```

Comment:
Here you've correctly removed `commit` and added a comment about it, but once it's gone, we don't know why you're talking about committing to conntrack :)
I think you can do it in a separate commit and have the commit message say `Return traffic should not be committed to conntrack.` .

Reply:
done